### ISSUE TITLE
GM_xmlhttpRequest limited per browser/script manager

### ISSUE THREAD

**Issue by trumpeta** [Tampermonkey/tampermonkey#1040]
(Please fill out the issue template with your details)

## Expected Behavior
I expect it gives on all supported browsers same result

## Actual Behavior
different result for same request on different browsers. See remarks at bottom

## Specifications

- browser: FF 81.0, Opera 71.0.3770.171, Chrome 85.0.4183.121, FF 54.0.1
- TM: 4.11.6117
- OS: Windows 8.1

## Remarks

It seems that some features of **GM_xmlhttpRequest** are blocked on modern browsers.
Specifically
- onreadystatechange callback, if assigned
- fetch param

I've written several scripts using GM_xmlhttpRequest, and have to rely on it as internal fetch() and XMLHttpRequest interfaces are blocked outide due CORS policy. So far for simple requests it works same on all browsers, but as the request is more structured, it seems the function has limited functionality on modern browsers. Where I get it fully functional is only Firefox 54 Legacy.
Since I have no evidence these two properties are ignored/blocked/failing, by debugging the request on FF Legacy, I easily get evidence that using the fetch=true attribute makes difference between response.status 200 and 400. The HTTP OK is only achieved with fetch parameter as true. So I use the same code on FF Quantum, and get HTTP/400 anyway. This makes me assume the parameter is not in effect.
The same is valid about onreadystatechange, whatever callback I hook to it and whatever the request is long, the callback is not even once called. Using same code, on FF Legacy, the callback is called in regular short intervals.

There's one more param, which seems not in effect on any browser, cookie. As there's no documentation about it's format, I've tried to assign it an object { name1: 'value', name2: 'value', ...etc } , a string with colon, a string with =, with same result. While I've examined by debugging the request in network inspector, that using and not using the cookie makes difference between OK and error status.

I feel this important function needs to be fixed in many ways.

**Comment by derjanb** [comment]
A first test showed that everything seems to work as expected here.

Can you please add an example script or make the attached one "fail"?

![image](https://user-images.githubusercontent.com/767504/94472616-47d40c00-01cb-11eb-9092-1b738f76bd0c.png)
![image](https://user-images.githubusercontent.com/767504/94472966-d6e12400-01cb-11eb-94e6-97931efc28a0.png)


```JS
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        *://example.com/*
// @grant        GM_xmlhttpRequest
// @connect      telecom-paristech.fr
// ==/UserScript==

GM_xmlhttpRequest({
  method: 'GET',
  url: 'http://download.tsi.telecom-paristech.fr/gpac/mp4box.js/mp4/h264bl.mp4?t=' + Math.random(),
  headers: {
    exmaple: 'testxhr',
  },
  cookie: 'bar=xhr' + Math.random(),
  overrideMimeType: 'application/octet-stream',
  responseType: 'arraybuffer',
  onreadystatechange(res) {
    console.log('xhr: onreadystatechange', JSON.stringify(res), this.response);
  },
  onload(res) {
    console.log('xhr: onload', res, this.response);
  },
  onloadstart(res) {
    console.log('xhr: onloadstart', res, this.response);
  },
  onerror(res) {
    console.log('xhr: onerror', res, this.response);
  },
  onprogress(res) {
    console.log('xhr: onprogress', res, this.response && this.response.byteLength, this.response);
  },
})


GM_xmlhttpRequest({
  method: 'GET',
  fetch: true,
  url: 'http://download.tsi.telecom-paristech.fr/gpac/mp4box.js/mp4/h264bl.mp4?t=' + Math.random(),
  headers: {
    exmaple: 'testfetch',
  },
  cookie: 'bar=fetch' + Math.random(),
  responseType: 'arraybuffer',
  fetch: 'fetch',
  onreadystatechange(res) {
    console.log('fetch: onreadystatechange', res, this.response && this.response.byteLength, this.response);
  },
  onload(res) {
    console.log('fetch: onload', res, this.response);
  },
  onloadstart(res) {
    console.log('fetch: onloadstart', res, this.response);
  },
  onerror(res) {
    console.log('fetch: onerror', res, this.response);
  },
  onprogress(res) {
    console.log('fetch: onprogress', res, this.response && this.response.byteLength, this.response);
  },
})```

**Comment by trumpeta** [comment]
It's pitty that unlike browser TM doesn't provide a diagnostic dump of own HTTP communication. In a snap,

```
		GM_xmlhttpRequest({ method: 'POST', url: 'https://ezgif.com/optipng',
		  headers: {
			'Accept': 'text/html',
			'Referer': 'https://ezgif.com/optipng',
			'Content-Type': 'multipart/form-data; boundary=' + boundary,
			'Content-Length': formData.length,
			'X-Requested-With': 'XMLHttpRequest',
		  },
		  data: formData,
		  fetch: true,
		  binary: true,
		  onload: response => {
			if (response.status < 200 || response.status >= 400) return reject(defaultErrorHandler(response));
			....
		  },
		  onerror: response => { reject(defaultErrorHandler(response)) },
		  ontimeout: response => { reject(defaultTimeoutHandler(response)) },
		});

```

Luckily I've captured the post header and body of two browsers which differ in the result for this request, where post data is bitwise same, except boundaries, and headers have only minor differences =>

**FF Legacy (receives "ok")**

```
Internet Protocol Version 4, Src: 192.168.137.2 (192.168.137.2), Dst: s1.ezgif.com (78.46.40.208)
Transmission Control Protocol, Src Port: 59836 (59836), Dst Port: https (443), Seq: 732, Ack: 151, Len: 467
Transport Layer Security
    TLSv1.2 Record Layer: Application Data Protocol: http2
        Content Type: Application Data (23)
        Version: TLS 1.2 (0x0303)
        Length: 462
        Encrypted Application Data: 000000000000000271c3f1bcb416df2431f4a09aaf103014…
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 13, Length 416, POST /optipng
        Length: 416
        Type: HEADERS (1)
        Flags: 0x24
        0... .... .... .... .... .... .... .... = Reserved: 0x0
        .000 0000 0000 0000 0000 0000 0000 1101 = Stream Identifier: 13
        [Pad Length: 0]
        0... .... .... .... .... .... .... .... = Exclusive: False
        .000 0000 0000 0000 0000 0000 0000 1011 = Stream Dependency: 11
        Weight: 21
        [Weight real: 22]
        Header Block Fragment: 83058660f5a4d5d53741872fb98d2ae43d3f877abbd07f66…
        [Header Length: 733]
        [Header Count: 16]
        Header: :method: POST
        Header: :path: /optipng
        Header: :authority: ezgif.com
        Header: :scheme: https
        Header: user-agent: Mozilla/5.0 (Windows NT 6.3; Win64; x64; rv:54.0) Gecko/20100101 Firefox/54.0
        Header: accept: text/html
        Header: accept-language: cs,en-GB;q=0.8,en-US;q=0.5,en;q=0.3
        Header: accept-encoding: gzip, deflate, br
        Header: tm_f8324referer: aHR0cHM6Ly9lemdpZi5jb20vb3B0aXBuZw==
        Header: content-type: multipart/form-data; boundary=--------WebKitFormBoundary-174D59F06C8
        Header: tm_f8324content-length: NzMxMTM1
        Header: x-requested-with: XMLHttpRequest
        Header: tm_f8324internal: ODQwZDBkZmItNGM1Yi00ZGJiLThjMDItYmYwNjUxMTQ4ZDQw
        Header: origin: moz-extension://1094f828-4a9a-495d-b4c3-740253b1bb44
        Header: content-length: 731135
        Header: cookie: cookieconsent_dismissed=yes
    Stream: WINDOW_UPDATE, Stream ID: 13, Length 4
        Length: 4
        Type: WINDOW_UPDATE (8)
        Flags: 0x00
        0... .... .... .... .... .... .... .... = Reserved: 0x0
        .000 0000 0000 0000 0000 0000 0000 1101 = Stream Identifier: 13
        0... .... .... .... .... .... .... .... = Reserved: 0x0
        .000 0000 1011 1110 0000 0000 0000 0000 = Window Size Increment: 12451840

```

**FF Quantum (receives "bad request")**

```
Internet Protocol Version 4, Src: 192.168.137.2 (192.168.137.2), Dst: s4.ezgif.com (116.202.237.233)
Transmission Control Protocol, Src Port: 32949 (32949), Dst Port: https (443), Seq: 768, Ack: 3948, Len: 372
Transport Layer Security
    TLSv1.3 Record Layer: Application Data Protocol: http2
        Opaque Type: Application Data (23)
        Version: TLS 1.2 (0x0303)
        Length: 367
        [Content Type: Application Data (23)]
        Encrypted Application Data: 6349c04b9478807151cd8326be97c57b610d91b3b67ab5c0…
HyperText Transfer Protocol 2
    Stream: HEADERS, Stream ID: 15, Length 328, POST /optipng
        Length: 328
        Type: HEADERS (1)
        Flags: 0x24
        0... .... .... .... .... .... .... .... = Reserved: 0x0
        .000 0000 0000 0000 0000 0000 0000 1111 = Stream Identifier: 15
        [Pad Length: 0]
        0... .... .... .... .... .... .... .... = Exclusive: False
        .000 0000 0000 0000 0000 0000 0000 0111 = Stream Dependency: 7
        Weight: 21
        [Weight real: 22]
        Header Block Fragment: 83058660f5a4d5d53741872fb98d2ae43d3f877abad07f66…
        [Header Length: 634]
        [Header Count: 16]
        Header: :method: POST
        Header: :path: /optipng
        Header: :authority: ezgif.com
        Header: :scheme: https
        Header: user-agent: Mozilla/5.0 (Windows NT 6.3; Win64; x64; rv:81.0) Gecko/20100101 Firefox/81.0
        Header: accept: text/html
        Header: accept-language: cs,en-GB;q=0.8,en-US;q=0.5,en;q=0.3
        Header: accept-encoding: gzip, deflate, br
        Header: content-type: multipart/form-data; boundary=--------WebKitFormBoundary-174D5FBB489
        Header: x-requested-with: XMLHttpRequest
        Header: origin: moz-extension://3cbb711c-422e-4056-919d-17aab1f155b6
        Header: content-length: 731135
        Header: dnt: 1
        Header: cookie: cookieconsent_dismissed=yes
        Header: referer: https://ezgif.com/optipng
        Header: te: trailers
    Stream: WINDOW_UPDATE, Stream ID: 15, Length 4
        Length: 4
        Type: WINDOW_UPDATE (8)
        Flags: 0x00
        0... .... .... .... .... .... .... .... = Reserved: 0x0
        .000 0000 0000 0000 0000 0000 0000 1111 = Stream Identifier: 15
        0... .... .... .... .... .... .... .... = Reserved: 0x0
        .000 0000 1011 1110 0000 0000 0000 0000 = Window Size Increment: 12451840

```

**Comment by trumpeta** [comment]
The code sample demonstrating my issue here:

```
// ==UserScript==
// @name         XHR-ISSUE
// @version      0.1
// @match        *://example.com/*
// @grant        GM_xmlhttpRequest
// @connect      *
// ==/UserScript==

let url = 'https://ezgif.com/optipng';
let payLoad = new URLSearchParams({
  'new-image-url': 'https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png',
});
GM_xmlhttpRequest({ method: 'POST', url: url,
  headers: {
   'Accept': 'text/html',
   'Referer': url,
   'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
   'Content-Length': payLoad.toString().length,
   'X-Requested-With': 'XMLHttpRequest',
  },
  data: payLoad.toString(),
  fetch: true,
  //binary: true,
  onload: response => { console.log('onload:', response) },
  onprogress: response => { console.log('onprogress:', response) },
  onerror: response => { console.log('onerror:', response) },
  ontimeout: response => { console.log('ontimeout:', response) },
});
```

fetch: true gives HTTP/200 on FF Legacy:

![](https://i.ibb.co/dKX22wV/Snag-3eecd94b.png)

fetch: false gives HTTP/400 on FF Legacy:, but calls onprogress callbacks:

![](https://i.ibb.co/gvq04Vw/Snag-3eed241f.png)

fetch: true/false gives HTTP/400 on FF Quantum, and onprogress callbacks aren't invoked

![](https://i.ibb.co/NSLy0sN/Snag-3eed37c7.png)

Is the POST method disabling them?

**Comment by derjanb** [comment]
That is working at FF legacy because it does not allow to set unsafe XHR headers. In your case the "content-length" header makes the request fail. However this is a server and/or userscript issue. 

BTW: you can debug TM's network requests at about:debugging#/runtime/this-firefox -> "Inspect"

This is how to make the request look like in FF legacy for the server:
```JS
...
GM_xmlhttpRequest({ method: 'POST', url: url,
  headers: {
   'Accept': 'text/html',
   'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
   'X-Requested-With': 'XMLHttpRequest',
   'Origin': ''
  },
  fetch: true,
  ...
```

**Comment by trumpeta** [comment]
Great, by removing the Content-Size header and got 200 (I think you messed the versions as Quantum/Chrome were the browsers where it fails). Does it mean that using Content-Size is a bad practise? If so when use and when not?

**Comment by trumpeta** [comment]
All has been resolved for me (thanks), except the missing updates to onprogress callback, on today's browsers (Quantum, Opera, Chrome). The typically initialised params for GM_xmlhttpRequest call are:

```
method: 'POST'
url: {whatever}
headers: { 'Content-Type': 'multipart/form-data; boundary=' + boundary }
data: {largeString}
binary: true
onload: response => { ... }
onprogress: response => { ... }
onerror: response => { ... }
ontimeout: response => { ... }
```

The difference here is that on FF Leagcy the callback is called intensively during readyState == LOADING, while on all of up to date browsers, the callback is called only on readyState == LOADED. It's not running in fetch mode.

**Comment by derjanb** [comment]
Duplicate of #958
